### ISSUE TITLE
NotImplementedError when using and_then and or_else

### ISSUE THREAD

**Issue by htanata** [tomdalling/resonad#1]
I tried to follow this example on README:

```ruby
result = Resonad.Success(5)
  .and_then { |i| Resonad.Success(i + 1) }
  .and_then { |i| Resonad.Failure("buzz #{i}") }
  .and_then { |i| Resonad.Success(i + 1) }
  .error #=> "buzz 6"
```

I got this error:

```
/Users/foo/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/resonad-1.1.0/lib/resonad.rb:135:in `flat_map': should be implemented in subclass (NotImplementedError)
	from resonad.rb:4:in `<main>'
```

Seems like it's something to do with `alias_method` and subclassing. `or_else` has the same problem too.

Here's a test case:

```diff
diff --git a/spec/resonad_spec.rb b/spec/resonad_spec.rb
index 4f47631..02d8fe4 100644
--- a/spec/resonad_spec.rb
+++ b/spec/resonad_spec.rb
@@ -35,6 +35,11 @@ RSpec.describe Resonad do
       expect(result).to eq('hello moto')
     end

+    it "can be and_then'd" do
+      result = subject.and_then{ |value| value + ' moto' }
+      expect(result).to eq('hello moto')
+    end
+
     it "can not be flat_map_error'd" do
       result = subject.flat_map_error{ |error| fail('shouldnt run this') }
       expect(result).to be(subject)
```

And the result:

```
Failures:

  1) Resonad Success can be and_then'd
     Failure/Error: raise NotImplementedError, "should be implemented in subclass"

     NotImplementedError:
       should be implemented in subclass
     # ./lib/resonad.rb:135:in `flat_map'
     # ./spec/resonad_spec.rb:39:in `block (3 levels) in <top (required)>'
```

Adding `alias_method` to the subclass like this seems to fix the problem, but I'm not sure if this is a good solution.

```diff
diff --git a/lib/resonad.rb b/lib/resonad.rb
index d467228..4b23775 100644
--- a/lib/resonad.rb
+++ b/lib/resonad.rb
@@ -54,6 +54,7 @@ class Resonad
     def flat_map
       yield value
     end
+    alias_method :and_then, :flat_map

     def flat_map_error
       self
```

I'm using version 1.1.0.

**Comment by tomdalling** [comment]
Thank you very much for submitting this (very detailed) issue. It's fixed now in version 1.1.1.

**Comment by htanata** [comment]
Nice. Thanks!
